### PR TITLE
fix: input function does not allow tailing spaces

### LINKED ISSUE #147: Input function does not allow tailing spaces
```
# select '[0.24    ]'::vector(1);
ERROR:  invalid input syntax for type vector: "0.24    ]"
LINE 1: select '[0.24    ]'::vector(1);
               ^
# select '[0.24]'::vector(1);
 vector 
--------
 [0.24]
(1 row)
```

### PR DESCRIPTION
**Pull request by yihong0618**
this fix #147 

also add test case for it

passed CI can check: https://github.com/yihong0618/pgvector/actions/runs/5209060111

### CODE REVIEW

**Review by jkatz — COMMENTED** [review]
This aligns with current PostgreSQL behavior with arrays, e.g.:

```
postgres=# SELECT '{1   , 2   }'::int[];
 int4  
-------
 {1,2}
(1 row)
```

However, I'd recommend following the convention that PostgreSQL has with the `array_isspace` function for the character checking (which would have to be reimplemented, as it's static):

https://github.com/postgres/postgres/blob/378d73ef204d0dcbeab834d52478e8cb90578ab7/src/backend/utils/adt/arrayfuncs.c#L438

Here is an example of how it is used in the `array_in` code:

https://github.com/postgres/postgres/blob/378d73ef204d0dcbeab834d52478e8cb90578ab7/src/backend/utils/adt/arrayfuncs.c#L257

**Comment by yihong0618** [comment]
> This aligns with current PostgreSQL behavior with arrays, e.g.:
> 
> ```
> postgres=# SELECT '{1   , 2   }'::int[];
>  int4  
> -------
>  {1,2}
> (1 row)
> ```
> 
> However, I'd recommend following the convention that PostgreSQL has with the `array_isspace` function for the character checking (which would have to be reimplemented, as it's static):
> 
> https://github.com/postgres/postgres/blob/378d73ef204d0dcbeab834d52478e8cb90578ab7/src/backend/utils/adt/arrayfuncs.c#L438
> 
> Here is an example of how it is used in the `array_in` code:
> 
> https://github.com/postgres/postgres/blob/378d73ef204d0dcbeab834d52478e8cb90578ab7/src/backend/utils/adt/arrayfuncs.c#L257

thanks will use `array_isspace`

File: src/vector.c [hunk 1]
```diff
@@ -86,6 +86,24 @@ CheckElement(float value)
 				 errmsg("infinite value not allowed in vector")));
 }
 
+/*
+ * PostgreSQL has with the array_isspace function for the character checking which is reimplemented
+ * here, as it's static, source code:
+ * https://github.com/postgres/postgres/blob/378d73ef204d0dcbeab834d52478e8cb90578ab7/src/backend/utils/adt/arrayfuncs.c#L438
+ */
+static inline bool
+array_isspace(char ch)
```
Comment (ankane) [inline src/vector.c:95]: Let's name this `vector_isspace`

File: src/vector.c [hunk 2]
```diff
@@ -150,6 +168,13 @@ vector_in(PG_FUNCTION_ARGS)
 	char	   *stringEnd;
 	Vector	   *result;
 
+	/*
+	 * Note: we currently allow whitespace between, but not within,
+	 * dimension items.
```
Comment (ankane) [inline src/vector.c:173]: I'm not sure this comment makes sense for vectors, so let's remove it

File: src/vector.c [hunk 3]
```diff
@@ -171,6 +196,9 @@ vector_in(PG_FUNCTION_ARGS)
 		x[dim] = strtof(pt, &stringEnd);
 		CheckElement(x[dim]);
 		dim++;
+		// stringEnd is space is OK to support SELECT '[   1  ,       2     ]'::vector(2);
+		while (array_isspace((unsigned char)*stringEnd))
```
Comment (ankane) [inline src/vector.c:200]: No need to cast here or below

File: test/sql/input.sql [hunk 4]
```diff
@@ -1,5 +1,9 @@
 SELECT '[1,2,3]'::vector;
 SELECT '[-1,2,3]'::vector;
+SELECT '[1,2,3  ]'::vector;
+SELECT '[1,     2,    3  ]'::vector;
+SELECT '[1.   , 2, 3 ]'::vector(3);
+SELECT ' [1,2,3] '::vector(3);
```
Comment (ankane) [inline test/sql/input.sql:6]: I think we can combine this into a single test

**Review by ankane — COMMENTED** [review]
Hi @yihong0618, thanks for reporting and the PR! Looks great. Added a few comments inline (accidentally submitted the first review early).

File: test/sql/input.sql (hunk 4 above)
Comment (yihong0618) [reply test/sql/input.sql:6]: copy that

File: src/vector.c (hunk 3 above)
Comment (yihong0618) [reply src/vector.c:200]: seems we need that, this is handle `[1   , 2    ]` which has spaces between `1 and ,`

File: src/vector.c (hunk 2 above)
Comment (yihong0618) [reply src/vector.c:173]: copy that

File: src/vector.c (hunk 3 above)
Comment (ankane) [reply src/vector.c:200]: `(unsigned char)` shouldn't be needed. Also, I think it should come after the `if (stringEnd == pt)` check.

Comment (yihong0618) [reply src/vector.c]: copy that, thanks for the info~

**Comment by yihong0618** [comment]
@ankane thanks very much for the review, the comments has been fixed now, and one CI on ubuntu seems failed, 
I am not sure its flaky test will check: https://github.com/yihong0618/pgvector/actions/runs/5217584573/jobs/9417561780

**Comment by ankane** [comment]
Awesome, big thanks @yihong0618!

**Comment by yihong0618** [comment]
@ankane thanks for your great repo~

**Pull request merged**